tests/gtest/avifallocationtest.cc

```diff
@@ -0,0 +1,177 @@
+
+#include <array>
+#include <limits>
+#include <tuple>
+
+#include "avif/avif.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+using ::testing::Combine;
+using ::testing::Values;
+using ::testing::ValuesIn;
+
+namespace libavif {
+namespace {
+
+void TestAllocation(uint32_t width, uint32_t height, uint32_t depth,
+                    avifResult expected_result) {
+  // The format of the image and which planes are allocated should not matter.
+  // Test all combinations.
+  for (avifPixelFormat format :
+       {AVIF_PIXEL_FORMAT_NONE, AVIF_PIXEL_FORMAT_YUV444,
+        AVIF_PIXEL_FORMAT_YUV422, AVIF_PIXEL_FORMAT_YUV420,
+        AVIF_PIXEL_FORMAT_YUV400}) {
+    for (avifPlanesFlag planes :
+         {AVIF_PLANES_YUV, AVIF_PLANES_A, AVIF_PLANES_ALL}) {
+      testutil::AvifImagePtr image(avifImageCreateEmpty(), avifImageDestroy);
+      ASSERT_NE(image, nullptr);
+      image->width = width;
+      image->height = height;
+      image->depth = depth;
+      image->yuvFormat = format;
+      EXPECT_EQ(avifImageAllocatePlanes(image.get(), planes), expected_result);
+
+      // Make sure the actual plane pointers are consistent with the settings.
+      if (expected_result == AVIF_RESULT_OK &&
+          format != AVIF_PIXEL_FORMAT_NONE &&
+          (planes == AVIF_PLANES_YUV || planes == AVIF_PLANES_ALL)) {
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_Y], nullptr);
+      } else {
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_Y], nullptr);
+      }
+      if (expected_result == AVIF_RESULT_OK &&
+          format != AVIF_PIXEL_FORMAT_NONE &&
+          format != AVIF_PIXEL_FORMAT_YUV400 && (planes & AVIF_PLANES_YUV)) {
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_U], nullptr);
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_V], nullptr);
+      } else {
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_U], nullptr);
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_V], nullptr);
+      }
+      if (expected_result == AVIF_RESULT_OK && (planes & AVIF_PLANES_A)) {
+        EXPECT_NE(image->alphaPlane, nullptr);
+      } else {
+        EXPECT_EQ(image->alphaPlane, nullptr);
+      }
+    }
+  }
+}
+
+TEST(AllocationTest, MinimumValid) { TestAllocation(1, 1, 8, AVIF_RESULT_OK); }
+
+TEST(AllocationTest, MaximumValid) {
+  // 8 bits
+  TestAllocation(std::numeric_limits<typeof(avifImage::width)>::max(), 1, 8,
+                 AVIF_RESULT_OK);
+  TestAllocation(1, std::numeric_limits<typeof(avifImage::height)>::max(), 8,
+                 AVIF_RESULT_OK);
+  // 12 bits (impacts the width because avifImage stride is stored as uint32_t)
+  TestAllocation(std::numeric_limits<typeof(avifImage::width)>::max() / 2, 1,
+                 12, AVIF_RESULT_OK);
+  TestAllocation(1, std::numeric_limits<typeof(avifImage::height)>::max(), 12,
+                 AVIF_RESULT_OK);
+  // Some high number of bytes that malloc() accepts to allocate.
+  TestAllocation(1024 * 16, 1024 * 16, 12, AVIF_RESULT_OK);  // Up to 2 GB total
```

Comment:
IMPORTANT: Please test this in 32-bit builds, whose address space is 4 GB total.

Reply:
Done. It passed for these 1024 * 16, 1024 * 16 dimensions but not for `std::numeric_limits<typeof(avifImage::width)>::max()`.

For the record, how to build as 32 bits:
- `sudo apt-get install gcc-multilib g++-multilib`
- Add `-DCMAKE_CXX_FLAGS=-m32 -DCMAKE_C_FLAGS=-m32` to libavif and local dependencies

Reply:
Thank you. Could you add this info to the README.md file?

Reply:
I meant adding the 32-bit build info to README.md **in a separate PR**.